streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java

```
@@ -1142,7 +1147,7 @@ void retryWithBackoff(final Map<TaskId, Set<TopicPartition>> tasksToBeCreated) {
                     }
                 }
 
-                if (tasksToBeCreated.isEmpty()) {
+                if (tasksToBeCreated.isEmpty() || backoffTimeMs > rebalanceTimeoutMs) {
```

I'm not sure this is correct. Lets say we have 2 tasks. Currently,  once the first task is successful we will sleep for `backoffTimeMs`. This shouldn't happen, i.e., once the task was successful `backoffTimeMs` should be set back to the initial value and we shouldn't sleep before we attempt to create the next task.

We need to track how long we've been rebalancing in total and use that to determine if we should give up (rather than the `backoffTimeMs`

In the event that it times out i think we should probably throw an exception so we leave the group, the thread shuts down (releasing any acquired locks), and some other thread can pick up the work.

---

`MAX_POLL_INTERVAL_MS_CONFIG` is absolute max time till this thread is kicked out. So this can be the upper limit for `backoffTimeMs`. I am not sure how can we track total time this thread was re balancing.  Also not throwing an exception but carrying on is OK, because it would try to commit offsets and it will fail that time, because it is no longer the owner of these partitions and already kicked out of the group. So it will try to join again and get new partitions.

---

Sorry - i'm reading it wrong. I was failing to see that the backoff is done in the outer loop. Anyway, i still think the `backoffTimeMs` should be reset after each successful task creation. 

As for tracking the total rebalance time... we could just get the time at the start of `onPartitionsAssigned` and make sure that `MAX_POLL_INTERVAL_MS` hasn't passed since then.

Yes you probably could rely on the `CommitFailedException` to be thrown by poll. My concern is that we carry on for some reason and some tasks are never processed. In which case it would be better to find a way to bail early so they can be re-assigned.

---

> Anyway, i still think the backoffTimeMs should be reset after each successful task creation.

I agree -- we missed this in the original implementation -- should have been like this from beginning on.

---

Thinking it a bit more, I feel even after KIP-129 this issue will still persists as long as we create one task and grab its local state file lock at a time. So how about the following solution (I feel it is not very elegant but could be correct in general):

1. In the `onPartitionAssigned`, do not create the tasks (both active and standby) but only does the cleaning of suspended ones; and remember the tasks needed to be created.

2. Then in the stream's main loop, check if there are any tasks that need to be created; if there are try creating them all by first grab all the locks and then creating them. If failed to grabbing all the locks then release the already-grabbed ones and the continue to the next loop, in which `consumer.poll()` will be triggered again and then it will be notified if it has lost the ownership or not. The key idea here is that either all tasks are created or none should be created.

3. after KIP-129 is introduced, we need to think whether the above pattern could be improved with `producer.initTxn` cals.

WDYT?

---

anyways this could be in a separate PR just throwing it out for discussions @sjmittal @dguy @mjsax We can merge the PR as is if other reviewers could give a thumb up.

---

Sounds good to me.